feat(APP-290): Implement feature flags system

tyhonchik opened the pull request:
# Description

**Feature Flags System (APP-290):**
- Implemented a centralized feature flag system with FeatureFlagsProvider
- Added CMS-based feature flag overrides support
- Integrated feature flags into LayoutRoot for server-side resolution
- Updated DebugPanel to use feature flags instead of environment variables

## Type of Change

<!--- Please delete the options that are not relevant. -->

- [ ] Major: Breaking change (change that would cause existing functionality to not work as expected)
- [x] Minor: Feature (non-breaking change which adds new functionality)
- [ ] Patch: Enhancement (non-breaking change to an existing feature)
- [ ] Patch: Bug fix (non-breaking change which fixes an issue)

## Developer Checklist:

- [ ] Manually smoke tested the functionality in a preview or locally
- [ ] Confirmed there are no new warnings or errors in the browser console
- [ ] (For User Stories only) Double-checked that all Acceptance Criteria are satisfied
- [ ] Confirmed there are no new warnings on automated tests
- [ ] Merged and published any dependent changes in downstream modules
- [ ] Selected the correct base branch
- [ ] Commented the code in hard-to-understand areas
- [ ] Followed the code style guidelines of this project
- [ ] Reviewed that the Files Changed in Github’s UI reflect my intended changes
- [ ] Confirmed the pipeline checks are not failing

## Review Checklist:

- [ ] (For User Stories only) Tested in a preview or locally that all Acceptance Criteria are satisfied
- [ ] Confirmed that changes follow the code style guidelines of this project

Fabricevladimir commented on packages/web-app/src/hooks/useExistingToken.tsx:
```diff
@@ -0,0 +1,30 @@
+  isTokenMintable: boolean;
+};
+
+export const useExistingToken = (
```
Might I suggest a rename since we're not actually using an existing token here? This is merely checking if the DAO token is mintable.

sepehr2github replied:
Yes, But I wanted to have this hook as a placeholder for future needs

Fabricevladimir commented on packages/web-app/src/containers/settings/majorityVoting/index.tsx:
```diff
@@ -39,6 +36,11 @@ const MajorityVotingSettings: React.FC<IPluginSettings> = ({daoDetails}) => {
     daoDetails?.plugins?.[0]?.instanceAddress || ''
   );
 
+  const {isTokenMintable} = useExistingToken(
+    daoToken?.address,
+    daoDetails?.address
+  );
+
```
Not exactly requesting a change, but it would be nice if the `useDaoToken` hook included whether the token is mintable in the response. Thoughts?

Fabricevladimir commented on packages/web-app/src/containers/settings/majorityVoting/index.tsx:
```diff
@@ -99,7 +85,7 @@ const MajorityVotingSettings: React.FC<IPluginSettings> = ({daoDetails}) => {
                 {tokenSupply?.formatted} {daoToken?.symbol}
               </p>
-              {showMintOption && <Tag label={t('labels.mintable')} />}
+              {isTokenMintable && <Tag label={t('labels.mintable')} />}
```
Stray comment, explicitly saying it's not mintable might be a good idea since the user isn't conveyed this and won't know unless they've gone through a DAO where the token is mintable.

sepehr2github replied:
Didn't get what you need from me here?

Fabricevladimir reviewed (commented):
Thanks, left a couple of comments. Also, something looks fishy about the commits on this PR...

tyhonchik committed:
feat(APP-290): Implement feature flags system

milosh86 commented on src/shared/utils/featureFlags/featureFlags.config.ts:
```diff
@@ -0,0 +1,70 @@
+import type { FeatureFlagDefinition, FeatureFlagEnvironment } from './featureFlags.api';
+
+const KNOWN_ENVIRONMENTS: FeatureFlagEnvironment[] = ['local', 'preview', 'development', 'staging', 'production'];
```
Both .api and .config files define the same list, let's have 1.

milosh86 commented on src/shared/utils/featureFlags/featureFlags.config.ts:
```diff
@@ -0,0 +1,70 @@
+        name: 'Debug panel',
+        description: 'Enables the in-app debug panel for developers.',
+        defaultValue: false,
+        enabledEnvironments: {
```
Maybe rename just to `environments` or `env` ... `enabledEnvironments` imply it's the list of enabled flags

milosh86 reviewed (commented):
Hey, looks good, left a couple of minor comments.

milosh86 reviewed (dismissed):
Looks good 🙌

thekidnamedkd commented on src/modules/application/components/providers/providers.test.tsx:
```diff
@@ -66,4 +72,9 @@ describe('<Providers /> component', () => {
         expect(screen.getByTestId('dialog-root-mock')).toBeInTheDocument();
     });
+
+    it('wraps children with the FeatureFlagsProvider', () => {
```
petty but just to stay consistent:

```suggestion
    it('renders the FeatureFlagsProvider with children', () => {
```

tyhonchik replied:
fixed

thekidnamedkd commented on src/shared/utils/featureFlags/featureFlags.config.ts:
```diff
@@ -0,0 +1,68 @@
+
+    // Fail-safe: treat unknown/missing environment as production to avoid
+    // accidentally enabling flags in non-explicit environments.
+    return 'production';
```
is production the safest fallback?

tyhonchik replied:
yep
e.g.
a) for some reason `production` didn't resolve -> we got production config -> nothing bad happend
b) for some reason `production` didn't resolve -> we got local config -> it's bad

thekidnamedkd commented on src/shared/utils/featureFlags/featureFlags.cookies.ts:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        const result: FeatureFlagOverrides = {};
+        const validKeys: FeatureFlagKey[] = ['debugPanel', 'subDao'];
```
Don't we already establish the keys in the config?

tyhonchik replied:
changed to FEATURE_FLAG_DEFINITIONS mapping directly

thekidnamedkd reviewed (commented):
Really nice. Good job on the README. My concerns are:

- Utils at minimum should always have tests, we have been lax on components for awhile however
- Utils historically have always followed the class singleton pattern
- Utils haven't previously used this dot notated file naming instead of subdirectories

thekidnamedkd commented on src/shared/featureFlags/featureFlags.constants.ts:
```diff
@@ -0,0 +1,68 @@
+ * @see {@link FeatureFlagDefinition} for detailed structure
+ * @see {@link ICmsFeatureFlagsResponse} for CMS override format
+ */
+export const FEATURE_FLAG_DEFINITIONS: FeatureFlagDefinition[] = [
```
this is maybe better independently in a constants directory?

tyhonchik replied:
I'm not sure
I didn't find better place for that. basically it is default `config`

tyhonchik commented:
> Really nice. Good job on the README. My concerns are:
> 
> * Utils at minimum should always have tests, we have been lax on components for awhile however

Valid point, I'll add some most important tests

> * Utils historically have always followed the class singleton pattern

Yes, FeatureFlags is used as a singleton. On line 119, a single instance is created and exported as featureFlags, and the class itself isn't exported. All imports share the same instance via module caching, so it behaves like a singleton.

> * Utils haven't previously used this dot notated file naming instead of subdirectories

I checked the existing utils and didn't see subfolder patterns like types/, consts/, etc. So I kept everything in dot-notated files to match the current structure.

tyhonchik committed:
feat: Refactor feature flags implementation

- Moved feature flags related imports to a dedicated module.
- Created tests for the feature flags system and its components.
- Updated README and folder structure.

tyhonchik commented:
@thekidnamedkd @milosh86
I’ve refactored the feature flags into a dedicated `shared/featureFlags` module (instead of living under utils), since this is more of a self-contained domain module than a generic utility. The structure is now split into `featureFlags.ts` (service), `featureFlags.constants.ts` (definitions + shared constants), `utils/` (env + cookie helpers), and `providers/` (GitHub CMS provider).

On top of that, I added tests for each part.

Could you take a look and see if this structure feels cleaner and more in line with how we organise other shared modules?

thekidnamedkd reviewed (approved):
Looking good now. I think if we have further semantic/style guide notes we can adjust later as we use the work here, if need be.
